<a href="https://colab.research.google.com/github/GSelvakumar/ml/blob/main/pandas_essentials.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Pandas

*  High Performance, Easy to use open source library for Data Analysis
*  Creates tabular format of data from different sources like csv, json, database
*  Have utilities for descriptive statistics, aggregation, handling missing data
*  Database utilities like merge, join are available
*  Fast, Programmable & easy alternative to spreadsheets





In [93]:
import pandas as pd
import numpy as np


# Understanding Series & Dataframes

*  Series represents one column
*  Combine mulitple columns to create a table(i.e., Dataframe)



In [94]:
series1 = pd.Series(data = [1, 2, 3, 4, 5], index=list("abcde"))
series1

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [95]:
series2 = pd.Series(data = [11, 22, 33, 44, 55], index=list("abcde"))
series2

a    11
b    22
c    33
d    44
e    55
dtype: int64

*  creating a **DataFrame** from the above two series
*  Data corresponding to the same index belongs to the same row

In [96]:
df = pd.DataFrame({'A': series1, 'B': series2})
df

A   B
a  1  11
b  2  22
c  3  33
d  4  44
e  5  55

creating a random **DataFrame** using numpy and pandas

In [97]:
df1 = pd.DataFrame(data= np.random.randint(1, 10, size=(10, 10)), index= list('ABCDEFGHIJ'), columns= list('abcdefghij'))
df1

a  b  c  d  e  f  g  h  i  j
A  3  9  1  5  2  4  1  8  7  7
B  9  3  7  5  4  4  4  5  5  2
C  4  8  5  2  9  1  5  7  8  9
D  9  3  7  8  3  6  8  2  1  6
E  9  2  1  5  8  2  4  3  7  3
F  4  8  3  6  3  7  9  7  4  2
G  1  1  9  6  7  7  8  8  5  9
H  7  3  6  6  1  6  9  4  7  5
I  2  4  6  4  7  7  3  7  2  7
J  5  2  7  4  6  3  9  3  4  1

# Loading CSV, JSON data

In [98]:
hr_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt')

In [99]:
hr_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


read_csv has a parameter called chunksize. The parameter essentially means the number of rows to be read into a dataframe at any single time in order to fit into the local memory

In [100]:
hr_data_itr = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt', chunksize=5000)


In [101]:
for hr_data in hr_data_itr:
  print(hr_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     5000 non-null   float64
 1   last_evaluation        5000 non-null   float64
 2   number_project         5000 non-null   int64  
 3   average_montly_hours   5000 non-null   int64  
 4   time_spend_company     5000 non-null   int64  
 5   Work_accident          5000 non-null   int64  
 6   left                   5000 non-null   int64  
 7   promotion_last_5years  5000 non-null   int64  
 8   sales                  5000 non-null   object 
 9   salary                 5000 non-null   object 
dtypes: float64(2), int64(6), object(2)
memory usage: 390.8+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 5000 to 9999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 ----------

In [102]:
pd.read_json('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/movie.json.txt')

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo                4.0              3.0  ...         3.5           NaN
Scarface               4.5              1.5  ...         3.0           NaN
Raging Bull            3.0              1.0  ...         NaN           4.5
Goodfellas             4.5              2.0  ...         4.5           NaN
The Apartment          1.0              5.0  ...         1.0           1.5
Roman Holiday          NaN              4.5  ...         3.0           NaN

[6 rows x 8 columns]

#Descriptive Statistics

Pandas api's for understanding data


In [103]:
hr_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt')

In [104]:
hr_data.head()  #returns the first five rows

satisfaction_level  last_evaluation  ...  sales  salary
0                0.38             0.53  ...  sales     low
1                0.80             0.86  ...  sales  medium
2                0.11             0.88  ...  sales  medium
3                0.72             0.87  ...  sales     low
4                0.37             0.52  ...  sales     low

[5 rows x 10 columns]

In [105]:
hr_data.tail()  #returns the last five rows

satisfaction_level  last_evaluation  ...    sales  salary
14994                0.40             0.57  ...  support     low
14995                0.37             0.48  ...  support     low
14996                0.37             0.53  ...  support     low
14997                0.11             0.96  ...  support     low
14998                0.37             0.52  ...  support     low

[5 rows x 10 columns]

In [106]:
hr_data.info()  #returns the information about the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [107]:
hr_data.describe() #Generate descriptive statistics.

satisfaction_level  last_evaluation  ...          left  promotion_last_5years
count        14999.000000     14999.000000  ...  14999.000000           14999.000000
mean             0.612834         0.716102  ...      0.238083               0.021268
std              0.248631         0.171169  ...      0.425924               0.144281
min              0.090000         0.360000  ...      0.000000               0.000000
25%              0.440000         0.560000  ...      0.000000               0.000000
50%              0.640000         0.720000  ...      0.000000               0.000000
75%              0.820000         0.870000  ...      0.000000               0.000000
max              1.000000         1.000000  ...      1.000000               1.000000

[8 rows x 8 columns]

In [108]:
hr_data.salary.value_counts()

low       7316
medium    6446
high      1237
Name: salary, dtype: int64

- renaming the column name

In [109]:
hr_data.rename(columns={'sales':'department'},inplace=True)
hr_data.head()

satisfaction_level  last_evaluation  ...  department  salary
0                0.38             0.53  ...       sales     low
1                0.80             0.86  ...       sales  medium
2                0.11             0.88  ...       sales  medium
3                0.72             0.87  ...       sales     low
4                0.37             0.52  ...       sales     low

[5 rows x 10 columns]

In [110]:
hr_data.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'department', 'salary'],
      dtype='object')

In [111]:
hr_data[['satisfaction_level','last_evaluation','number_project']].head()

satisfaction_level  last_evaluation  number_project
0                0.38             0.53               2
1                0.80             0.86               5
2                0.11             0.88               7
3                0.72             0.87               5
4                0.37             0.52               2

In [112]:
hr_data.satisfaction_level[:5]

0    0.38
1    0.80
2    0.11
3    0.72
4    0.37
Name: satisfaction_level, dtype: float64

In [113]:
hr_data['satisfaction_level'][:5]

0    0.38
1    0.80
2    0.11
3    0.72
4    0.37
Name: satisfaction_level, dtype: float64

In [114]:
movie_data = pd.read_json('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/movie.json.txt')
movie_data

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo                4.0              3.0  ...         3.5           NaN
Scarface               4.5              1.5  ...         3.0           NaN
Raging Bull            3.0              1.0  ...         NaN           4.5
Goodfellas             4.5              2.0  ...         4.5           NaN
The Apartment          1.0              5.0  ...         1.0           1.5
Roman Holiday          NaN              4.5  ...         3.0           NaN

[6 rows x 8 columns]

# **accessing data using index values**

In [115]:
movie_data.loc['Scarface'] #access using index values

David Smith         4.5
Brenda Peterson     1.5
Bill Duffy          5.0
Samuel Miller       3.5
Julie Hammel        2.5
Clarissa Jackson    4.5
Adam Cohen          3.0
Chris Duncan        NaN
Name: Scarface, dtype: float64

In [116]:
movie_data.loc['Vertigo':'Raging Bull']

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo              4.0              3.0  ...         3.5           NaN
Scarface             4.5              1.5  ...         3.0           NaN
Raging Bull          3.0              1.0  ...         NaN           4.5

[3 rows x 8 columns]

In [117]:
movie_data['Vertigo':'Raging Bull']

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo              4.0              3.0  ...         3.5           NaN
Scarface             4.5              1.5  ...         3.0           NaN
Raging Bull          3.0              1.0  ...         NaN           4.5

[3 rows x 8 columns]

In [118]:
movie_data.iloc[1]  #access using index numbers

David Smith         4.5
Brenda Peterson     1.5
Bill Duffy          5.0
Samuel Miller       3.5
Julie Hammel        2.5
Clarissa Jackson    4.5
Adam Cohen          3.0
Chris Duncan        NaN
Name: Scarface, dtype: float64

In [119]:
movie_data.iloc[1:4]

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Scarface             4.5              1.5  ...         3.0           NaN
Raging Bull          3.0              1.0  ...         NaN           4.5
Goodfellas           4.5              2.0  ...         4.5           NaN

[3 rows x 8 columns]

In [120]:
movie_data[1:4]

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Scarface             4.5              1.5  ...         3.0           NaN
Raging Bull          3.0              1.0  ...         NaN           4.5
Goodfellas           4.5              2.0  ...         4.5           NaN

[3 rows x 8 columns]

filtering based on conditional statements

In [121]:
movie_data[ (movie_data['Adam Cohen'] > 3)]

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo             4.0              3.0  ...         3.5           NaN
Goodfellas          4.5              2.0  ...         4.5           NaN

[2 rows x 8 columns]

In [122]:
movie_data[ ((movie_data['Adam Cohen'] > 3) & (movie_data['David Smith'] > 4))] 

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Goodfellas          4.5              2.0  ...         4.5           NaN

[1 rows x 8 columns]

# Handling Missing data

- Machine learning algorithm do not accept data missing
- If there is a column with more than 40% of data missing then we may drop the column
- For rows with, important column values missing. Drop the rows

In [123]:
movie_data

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo                4.0              3.0  ...         3.5           NaN
Scarface               4.5              1.5  ...         3.0           NaN
Raging Bull            3.0              1.0  ...         NaN           4.5
Goodfellas             4.5              2.0  ...         4.5           NaN
The Apartment          1.0              5.0  ...         1.0           1.5
Roman Holiday          NaN              4.5  ...         3.0           NaN

[6 rows x 8 columns]

In [124]:
movie_data['Bill Duffy'].notnull()

Vertigo           True
Scarface          True
Raging Bull      False
Goodfellas        True
The Apartment     True
Roman Holiday    False
Name: Bill Duffy, dtype: bool

In [125]:
movie_data[movie_data['Bill Duffy'].notnull()]

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Vertigo                4.0              3.0  ...         3.5           NaN
Scarface               4.5              1.5  ...         3.0           NaN
Goodfellas             4.5              2.0  ...         4.5           NaN
The Apartment          1.0              5.0  ...         1.0           1.5

[4 rows x 8 columns]

In [126]:
movie_data[movie_data['Bill Duffy'].isnull()]  #get all the rows which bill duffy is null

David Smith  Brenda Peterson  ...  Adam Cohen  Chris Duncan
Raging Bull            3.0              1.0  ...         NaN           4.5
Roman Holiday          NaN              4.5  ...         3.0           NaN

[2 rows x 8 columns]

# Dropping Rows and Columns


In [127]:
titanic_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/titanic-train.csv.txt')

In [128]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Dropping 'Cabin' column as it has only 204 data present in 891 rows

In [129]:
titanic_data.drop(['Cabin'],axis=1,inplace=True)

if inplace is true then it returns none, then when inplace is false returns a copy of the object with the operation performed.

In [130]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


- Now drop all rows with missing values
- We don't have inplace=True, so doesn't modify the dataframe 


In [131]:
titanic_data.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


- Consider only selected columns to check if they contain NA

In [132]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


[dropna has various types of removal of missing data](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html)

- here dropna will remove the rows even with one missing element in it

In [133]:
titanic_data.dropna(subset=['Embarked','Age']).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


In [134]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


Another approach of handling missing data is filling the missing ones

In [135]:
titanic_data.fillna({'Age':0,'Embarked':'Unknown'}).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [136]:
titanic_data.Age.fillna(method='ffill')[:5]

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

# Handling Duplicates

- Sometimes, it is difficult to ensure that data is not duplicated
- This becomes responsibility in Data cleaning step to make sure that duplicated data is deleted.

In [137]:
df2 = pd.DataFrame({'A':[1,1,3,4,5,1], 'B':[1,1,3,7,8,1], 'C':[3,1,1,6,7,1]})

In [138]:
df2

A  B  C
0  1  1  3
1  1  1  1
2  3  3  1
3  4  7  6
4  5  8  7
5  1  1  1

In [139]:
df2.duplicated()

0    False
1    False
2    False
3    False
4    False
5     True
dtype: bool

In [140]:
df2[df2.duplicated()]

A  B  C
5  1  1  1

- subset is default of none. To verify only specific columns use subset and mention that column names in list as the input to that attribute.

[Only consider certain columns for identifying duplicates, by default use all of the columns.](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html?highlight=subset%20duplicated)



In [141]:
df[df.duplicated(subset=['A','B'])]

Empty DataFrame
Columns: [A, B]
Index: []

# Function Application

- map for transforming one column to another
- can be applied only to series
- rolling windows

In [142]:
titanic_data.head()

PassengerId  Survived  Pclass  ...            Ticket     Fare  Embarked
0            1         0       3  ...         A/5 21171   7.2500         S
1            2         1       1  ...          PC 17599  71.2833         C
2            3         1       3  ...  STON/O2. 3101282   7.9250         S
3            4         1       1  ...            113803  53.1000         S
4            5         0       3  ...            373450   8.0500         S

[5 rows x 11 columns]

In [143]:
titanic_data_age = titanic_data[titanic_data.Age.notnull()]

creating a new column called **age_category** which is mapped using the lambda function.

In [144]:
titanic_data['age_category'] = titanic_data.Age.map(lambda age: 'Kid' if age < 18 else 'Adult')

In [145]:
titanic_data.head()

PassengerId  Survived  Pclass  ...     Fare Embarked  age_category
0            1         0       3  ...   7.2500        S         Adult
1            2         1       1  ...  71.2833        C         Adult
2            3         1       3  ...   7.9250        S         Adult
3            4         1       1  ...  53.1000        S         Adult
4            5         0       3  ...   8.0500        S         Adult

[5 rows x 12 columns]

In [146]:
#apply can be used in series as well as DataFrame
titanic_data.Age.apply('sum')

21205.17

returns the first 15 manipulated data of age

- apply on dataframes helps us dealing with multiple columns
- func will receive all the rows

In [147]:
titanic_data.Age.apply(lambda age: 'Kid' if age < 18 else 'Adult')[:15]

0     Adult
1     Adult
2     Adult
3     Adult
4     Adult
5     Adult
6     Adult
7       Kid
8     Adult
9       Kid
10      Kid
11    Adult
12    Adult
13    Adult
14      Kid
Name: Age, dtype: object

In [148]:
#e will be each row

def func(e):
    if e.Sex == 'male':
        return e.Fare * 2
    else:
        return e.Fare

In [149]:
titanic_data.apply(func,axis=1)[:5]

0    14.5000
1    71.2833
2     7.9250
3    53.1000
4    16.1000
dtype: float64

In [150]:
titanic_data[:10]

PassengerId  Survived  Pclass  ...     Fare Embarked  age_category
0            1         0       3  ...   7.2500        S         Adult
1            2         1       1  ...  71.2833        C         Adult
2            3         1       3  ...   7.9250        S         Adult
3            4         1       1  ...  53.1000        S         Adult
4            5         0       3  ...   8.0500        S         Adult
5            6         0       3  ...   8.4583        Q         Adult
6            7         0       1  ...  51.8625        S         Adult
7            8         0       3  ...  21.0750        S           Kid
8            9         1       3  ...  11.1333        S         Adult
9           10         1       2  ...  30.0708        C           Kid

[10 rows x 12 columns]

groupby - It splits data into groups, a function is applied to each groups separately, combine results into a data structure

In [151]:
titanic_data.groupby(['Sex']).Age.mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [152]:
titanic_data.groupby(['Sex']).Age.agg(['mean','min','max'])

mean   min   max
Sex                          
female  27.915709  0.75  63.0
male    30.726645  0.42  80.0

In [153]:
titanic_data.Age.rolling(window=5,min_periods=1).agg(['sum','min'])[:10]

sum   min
0   22.0  22.0
1   60.0  22.0
2   86.0  22.0
3  121.0  22.0
4  156.0  22.0
5  134.0  26.0
6  150.0  26.0
7  126.0   2.0
8  118.0   2.0
9   97.0   2.0

The **rolling** is used to split the dataset and perform statistics. The window size is the defines the set of calculation performs.


```
titanic_data.Age.rolling(window=5, min_periods=1).sum()
```

here the window size is 5 so it will perform cumulative product for the index of 0 to 4 then for the index 5 it will give the cumulative product result of index 2 to 5, this will happens for every 5 index set.

---

eg:

- 0 | 22 |  22
- 1 | 38 |  60   (22 + 38)
- 2 | 26 |  86   (60 + 26)
- 3 | 35 |  121  (86 + 35)
- 4 | 35 |  155  (121  35)
- 5 | 54 |  188  (38 + 26 + 35 + 35 + 54) 

from index 2 to 5 (here index 5 is not present in data set)










In [154]:
titanic_data[titanic_data.Name.str.contains('Mr')]

PassengerId  Survived  Pclass  ...     Fare Embarked  age_category
0              1         0       3  ...   7.2500        S         Adult
1              2         1       1  ...  71.2833        C         Adult
3              4         1       1  ...  53.1000        S         Adult
4              5         0       3  ...   8.0500        S         Adult
5              6         0       3  ...   8.4583        Q         Adult
..           ...       ...     ...  ...      ...      ...           ...
883          884         0       2  ...  10.5000        S         Adult
884          885         0       3  ...   7.0500        S         Adult
885          886         0       3  ...  29.1250        Q         Adult
889          890         1       1  ...  30.0000        C         Adult
890          891         0       3  ...   7.7500        Q         Adult

[647 rows x 12 columns]

# Append, Merge, Join & Concatenate

- Append for stacking dataframe

In [157]:
df1 = pd.DataFrame(data=np.random.randint(1,10,size=(10,3)), columns=list('ABC'))
df1

A  B  C
0  2  5  6
1  5  8  6
2  7  9  7
3  5  9  8
4  8  8  7
5  7  9  5
6  7  7  5
7  7  8  4
8  3  6  1
9  1  4  2

In [158]:
df2 = pd.DataFrame(data=np.random.randint(1,10,size=(10,3)), columns=list('ABC'))
df2

A  B  C
0  4  4  3
1  9  7  2
2  4  6  2
3  9  8  4
4  5  9  7
5  3  7  9
6  6  8  9
7  2  1  1
8  5  2  3
9  1  6  8

append will add the second dataframe below the first dataframe. The index value will continue if we mentioned the ignore_index = True else the index of the second dataframe will be append without any change like

df1 - 0, 1, 2,.., 8, 9 then the df2 index as 0, 1, 2, ...

In [161]:
df1.append(df2, ignore_index=True)

A  B  C
0   2  5  6
1   5  8  6
2   7  9  7
3   5  9  8
4   8  8  7
5   7  9  5
6   7  7  5
7   7  8  4
8   3  6  1
9   1  4  2
10  4  4  3
11  9  7  2
12  4  6  2
13  9  8  4
14  5  9  7
15  3  7  9
16  6  8  9
17  2  1  1
18  5  2  3
19  1  6  8

merge will takes place based on the order and key parameter mentioned. It will only merge the the values which has common key in both the dataframe

In [162]:
left = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3','K4','K5'],
                         'A': ['A0', 'A1', 'A2', 'A3','A4','A5'],
                         'B': ['B0', 'B1', 'B2', 'B3','B4','B5']})

right = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3','K6','K7'],
                          'C': ['C0', 'C1', 'C2', 'C3','C6','C7'],
                         'D': ['D0', 'D1', 'D2', 'D3','D6','D7']})

In [163]:
left.merge(right, on='key')

key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K1  A1  B1  C1  D1
2  K2  A2  B2  C2  D2
3  K3  A3  B3  C3  D3

In [164]:
right.merge(left, on='key')


key   C   D   A   B
0  K0  C0  D0  A0  B0
1  K1  C1  D1  A1  B1
2  K2  C2  D2  A2  B2
3  K3  C3  D3  A3  B3

In [165]:
left.merge(right, on='key', how='left')

key   A   B    C    D
0  K0  A0  B0   C0   D0
1  K1  A1  B1   C1   D1
2  K2  A2  B2   C2   D2
3  K3  A3  B3   C3   D3
4  K4  A4  B4  NaN  NaN
5  K5  A5  B5  NaN  NaN

Join is for combining data based on index values

In [166]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'],
                     'B': ['B0', 'B1', 'B2']},
                     index=['K0', 'K1', 'K2'])

right = pd.DataFrame({'C': ['C0', 'C2', 'C3'],
                      'D': ['D0', 'D2', 'D3']},
                     index=['K0', 'K2', 'K3'])

In [167]:
left.join(right)

A   B    C    D
K0  A0  B0   C0   D0
K1  A1  B1  NaN  NaN
K2  A2  B2   C2   D2

# Pivot tables

It is the useful way to get more information about the data


In [169]:
sales_data = pd.read_excel('https://github.com/zekelabs/data-science-complete-tutorial/blob/master/Data/sales-funnel.xlsx?raw=true')
sales_data

Account                          Name  ...   Price     Status
0    714466               Trantow-Barrows  ...   30000  presented
1    714466               Trantow-Barrows  ...   10000  presented
2    714466               Trantow-Barrows  ...    5000    pending
3    737550  Fritsch, Russel and Anderson  ...   35000   declined
4    146832                  Kiehn-Spinka  ...   65000        won
5    218895                     Kulas Inc  ...   40000    pending
6    218895                     Kulas Inc  ...   10000  presented
7    412290                 Jerde-Hilpert  ...    5000    pending
8    740150                    Barton LLC  ...   35000   declined
9    141962                    Herman LLC  ...   65000        won
10   163416                   Purdy-Kunde  ...   30000  presented
11   239344                    Stokes LLC  ...    5000    pending
12   239344                    Stokes LLC  ...   10000  presented
13   307599   Kassulke, Ondricka and Metz  ...    7000        won
14   688981                   Keeling LLC  ...  100000        won
15   729833                     Koepp Ltd  ...   65000   declined
16   729833                     Koepp Ltd  ...    5000  presented

[17 rows x 8 columns]

In [170]:
pd.pivot_table(sales_data, index=['Manager','Rep'], values=['Account','Price'], aggfunc=[np.sum, np.mean])

sum              mean              
                             Account   Price   Account         Price
Manager       Rep                                                   
Debra Henley  Craig Booker   2880948   80000  720237.0  20000.000000
              Daniel Hilton   584622  115000  194874.0  38333.333333
              John Smith     1152440   40000  576220.0  20000.000000
Fred Anderson Cedric Moss     784066  110000  196016.5  27500.000000
              Wendy Yule     2456246  177000  614061.5  44250.000000

# Normalising JSON

- JSON data will not always be of flat but can be hierchial


In [171]:
data = [     {'state': 'Florida',
              'shortname': 'FL',
              'info': {
                   'governor': 'Rick Scott'
              },
              'counties': [{'name': 'Dade', 'population': 12345},
                          {'name': 'Broward', 'population': 40000},
                          {'name': 'Palm Beach', 'population': 60000}]},
             {'state': 'Ohio',
              'shortname': 'OH',
              'info': {
                   'governor': 'John Kasich'
              },
              'counties': [{'name': 'Summit', 'population': 1234},
                           {'name': 'Cuyahoga', 'population': 1337}]}]

In [172]:
from pandas.io.json import json_normalize

In [173]:
json_normalize(data)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """Entry point for launching an IPython kernel.


state  ... info.governor
0  Florida  ...    Rick Scott
1     Ohio  ...   John Kasich

[2 rows x 4 columns]

In [174]:
json_normalize(data,'counties',['state',['info', 'governor']])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """Entry point for launching an IPython kernel.


name  population    state info.governor
0        Dade       12345  Florida    Rick Scott
1     Broward       40000  Florida    Rick Scott
2  Palm Beach       60000  Florida    Rick Scott
3      Summit        1234     Ohio   John Kasich
4    Cuyahoga        1337     Ohio   John Kasich